src/System.Private.CoreLib/shared/System/IO/FileStream.Windows.cs

```diff
@@ -226,47 +226,68 @@ protected override void Dispose(bool disposing)
-                    if (_writePos > 0)
+                    try
                     {
-                        try
-                        {
-                            FlushWriteBuffer(!disposing);
-                        }
-                        catch (Exception e) when (IsIoRelatedException(e) && !disposing)
-                        {
-                            // On finalization, ignore failures from trying to flush the write buffer,
-                            // e.g. if this stream is wrapping a pipe and the pipe is now broken.
-                        }
+                        FlushWriteBuffer(!disposing);
+                    }
+                    catch (Exception e) when (IsIoRelatedException(e) && !disposing)
+                    {
+                        // On finalization, ignore failures from trying to flush the write buffer,
+                        // e.g. if this stream is wrapping a pipe and the pipe is now broken.
                     }
                 }
             }
             finally
             {
                 if (_fileHandle != null && !_fileHandle.IsClosed)
                 {
-                    if (_fileHandle.ThreadPoolBinding != null)
-                        _fileHandle.ThreadPoolBinding.Dispose();
-
+                    _fileHandle.ThreadPoolBinding?.Dispose();
                     _fileHandle.Dispose();
                 }
 
-                if (_preallocatedOverlapped != null)
-                    _preallocatedOverlapped.Dispose();
-
+                _preallocatedOverlapped?.Dispose();
                 _canSeek = false;
 
                 // Don't set the buffer to null, to avoid a NullReferenceException
                 // when users have a race condition in their code (i.e. they call
                 // Close when calling another method on Stream like Read).
-                //_buffer = null;
-                base.Dispose(disposing);
+            }
+        }
+
+        public override ValueTask DisposeAsync() =>
+            GetType() == typeof(FileStream) ?
+                DisposeAsyncCore() :
+                base.DisposeAsync();
+
+        private async ValueTask DisposeAsyncCore()
+        {
+            // Same logic as in Dispose(), except with async counterparts.
+            // TODO: https://github.com/dotnet/corefx/issues/32837: FlushAsync does synchronous work.
+            try
+            {
+                if (_fileHandle != null && !_fileHandle.IsClosed && _writePos > 0)
+                {
+                    await FlushAsyncInternal(default).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                if (_fileHandle != null && !_fileHandle.IsClosed)
+                {
+                    _fileHandle.ThreadPoolBinding?.Dispose();
+                    _fileHandle.Dispose();
+                }
+
+                _preallocatedOverlapped?.Dispose();
+                _canSeek = false;
+                GC.SuppressFinalize(this);
```

Comment:
GC.SuppressFinalize shouldn't be in the finally... if FlushAsyncInternal throws then SuppressFinalize should not be called.

Reply:
If we get here, we will have already closed the file handle, at which point Dispose(false) as called from the finalizer is a nop.

Reply:
> If we get here, we will have already closed the file handle

Okay, so that's a little fishy; but is definitely what the synchronous Dispose is doing.  So, I guess, just add a comment? :smile:

Reply:
Not to change anything here, but in a perfect world, what would you want it to do? Having the finalizer retry the flush seems wrong, at which point the finalizer would just be closing the handle, so why not close it in the dispose and then suppress the finalization?

Reply:
Really I'm only thinking about it from having seen the section in Framework Design Guidelines; which is, essentially, "maybe the thing that made Dispose throw was transient, and it will get a second chance in Finalize", which is why the recommendation for Dispose(void) is `Dispose(true); GC.SuppressFinalize(this);` instead of the other way around.